text/0000-post-installation-mirror.md

```diff
@@ -0,0 +1,124 @@
+- Download additional dependencies or codes
+- Compile node native extensions
+
+This observation leads to the assumption that most installation scripts only modify files and directories within its own folders.  Consequently, we can store the post installation content as artifacts without worrying about inter-module dependencies.
+
+##Modification to `yarn` offline mirror structure and `yarn.lock`
+
+- Store post installation artifacts under post-installation subdirectory when using yarn offline mirror.  _resolved_ field should reflect this change as well.
+
+examples:
+```
+node-pre-gyp@^0.6.29:
+  version "0.6.31"
+  resolved "post-installation/node-pre-gyp-0.6.31.tgz#sha1"
+  dependencies:
+    mkdirp "~0.5.1"
+    nopt "~3.0.6"
+    npmlog "^4.0.0"
+    ...
+```
+
+##Modification to cli
+
+Storing post installation artifacts should be a voluntary feature.  This means that we should only store post installation artifacts when asked specifically.  When installing from offline mirror, post installation artifacts should be given preference since the existence of such artifacts reflects specific actions taken by the maintainer of the offline mirror previously.
+
+- For _add_ command, add an additional command line parameter --save-post-install. When this parameter is specified, store post installation artifact to offline mirror. The post install artifact should include all files and directories after running installation scripts, but exclude node_modules subdirectory.
+
+- For _install_ command, search post install subdirectory first, if there is no post installation artifacts, fail through to existing work flow.
+
+- When installation a post install artifact from offline mirror
+  - extract the tar.gz file in place
+  - do not run install scripts
+  - install dependencies and create bin link as usual
+
+# How We Teach This
+
+*What names and terminology work best for these concepts and why?*
+"post installation artifacts" to distinguish from node modules
+"network restricted environments (NRE)" where access to Internet is controlled
+"hermetic builds" means all dependencies are included, could be used as synonym for "deterministic builds"
+
+*How is this idea best presented?*
+
+- Emphases on the concept of "hermetic builds".
+- Emphases on `yarn` usage within "network restricted environments (NRE)".
+
+*Would the acceptance of this proposal mean the Yarn documentation must be re-organized or altered?*
+
+No.
+
+*Does it change how Yarn is taught to new users at any level?*
+
+This feature should be considered as advanced and should be taught specifically to following users:
+- users who need to operate within network restricted environments (NRE)
+- users who value deterministic dependency resolution and hermetic builds above convenience
+
+*How should this feature be introduced and taught to existing Yarn users?*
+Explain the intended use case with illustrated work flow.
+
+# Drawbacks

```

Comment:
1. Here is an example with node-gyp https://github.com/yarnpkg/yarn/issues/1874
2. Yep, I know the pain but we have to deal with it as many projects are sharing same yarn.lock files and offline mirror .tgz files across all OS
3. I am pretty sure a postinstall cache should be independent from offline mirror